packages/target-decisioning-engine/src/contextProvider.js

```diff
@@ -82,13 +82,19 @@ function createBrowserContext(context) {
  * @param { string } url
  * @return { import("../types/DecisioningContext").PageContext }
  */
-function createUrlContext(url) {
+function createUrlContext(url, parseURLImpl) {
```

Comment:
I looked at the two parseURL implementations and the new one shares a lot of code from the original.  The only real difference is domain parsing logic.  So instead I think we should extract the domain parsing logic from the original into a basic `parseDomainBasic` method.  And make a new method in `parsing.js` called `parseDomainFull` that uses the `psl` library.  That way we can avoid code duplication.

For example...
in decisioning-engine/utils.js:

```js
/**
 * @param {string} host
 * @returns {{subdomain?:string, domain?: string, topLevelDomain?: string}}
 */
export function parseDomainBasic(host) {
  const result = {};
  const domainParts = host.split(".");

  switch (domainParts.length) {
    case 1:
      result.subdomain = "";
      result.domain = host;
      result.topLevelDomain = "";
      break;
    case 2:
      result.subdomain = "";
      result.domain = host;
      result.topLevelDomain = domainParts[1];
      break;
    case 3:
      result.subdomain = domainParts[0] === "www" ? "" : domainParts[0];
      result.domain = host;
      result.topLevelDomain = domainParts[2];
      break;
    case 4:
      result.subdomain = domainParts[0] === "www" ? "" : domainParts[0];
      result.domain = host;
      result.topLevelDomain = `${domainParts[2]}.${domainParts[3]}`;
      break;
    default:
      break;
  }

  return result;
}

/**
 *
 * @param {string} url
 * @param { import("../types/DecisioningContext").ParseDomainFunc } parseDomain
 * @returns {{path: string, fragment: string, topLevelDomain?: string, query: string, domain?: string, subdomain?: string, url: string}}
 */
export function parseURL(url, parseDomain = parseDomainBasic) {
  if (!isString(url)) {
    // eslint-disable-next-line no-param-reassign
    url = "";
  }

  const parsed = parseURI(url) || {};

  const { host = "", path = "", query = "", anchor = "" } = parsed;

  return {
    url,
    path,
    query,
    fragment: anchor,
    ...parseDomain(host)
  };
}
```


and in target-tools/parsing.js you can reduce your method and rename it to something like this...

```js

/**
 * @param {string} host
 * @returns {import("../../target-decisioning-engine/types/DecisioningContext").DomainContext}
 */
export function parseDomainFull(host) {
  const result = {};

  if (typeof host !== "string" || !psl.isValid(host)) {
    return result;
  }

  const parseResult = psl.parse(host);

  if (parseResult.subdomain) {
    result.subdomain = parseResult.subdomain.startsWith("www")
      ? parseResult.subdomain.substring(4)
      : parseResult.subdomain;
  } else {
    result.subdomain = "";
  }

  result.domain = parseResult.domain;
  result.topLevelDomain = parseResult.tld;

  return result;
}
```

Reply:
In this sample code you'll notice I added some JSDOC.  Please be sure to add these headers for readability.  You can add interfaces to explain expected objects more easily like so:

in target-decisioning-engine/types/DecisioningContext.d.ts

```ts

export interface DomainContext {
  subdomain?: string;
  domain?: string;
  topLevelDomain?: string;
}

export interface ParseDomainFunc {
  (string): DomainContext;
}

```

☝️ these interfaces reveal (at least at developer and IDE level) the expected contract for the `parseDomain` method